synchronicity/genstub.py

```diff
@@ -0,0 +1,131 @@
+import inspect
+import types
+from unittest import mock
+
+class ReprObj:
+    def __init__(self, repr):
+        self._repr = repr
+
+    def __repr__(self):
+        return self._repr
+
+    def __str__(self):
+        return self._repr
+
+
+
+class StubEmitter:
+    def __init__(self, target_module):
+        self.target_module = target_module
+        self.imports = set()
+        self.parts = []
+        self._indentation = "    "
+
+
+    def formatannotation(self, annotation, base_module=None):
+        assert base_module is None  # don't think this arg is used by signature, but lets check
+        # modified version of the stdlib formatannotations:
+        # if isinstance(annotation, types.GenericAlias):
+        #     return str(annotation)
+        origin = getattr(annotation, "__origin__", None)
+        if origin is None:
+            if isinstance(annotation, type):
+                if annotation == None.__class__:
+                    return "None"
+                if annotation.__module__ in ('builtins', self.target_module):
+                    return annotation.__qualname__
+                return annotation.__module__+'.'+annotation.__qualname__
+            return repr(annotation)
+        # generic:
+        args = getattr(annotation, "__args__", ())
+        return annotation.copy_with(tuple(ReprObj(self.formatannotation(arg)) for arg in args))
+
+    def indent(self, level):
+        return level * self._indentation
+
+    def _get_function(self, func, indentation_level=0):
+        # return source code of function and track imports
+        for annotation in func.__annotations__.values():
+            self._type_imports(annotation)
+
+        return self._get_func_stub_source(func, indentation_level)
+
+    def add_function(self, func, indentation_level=0):
+        # adds function source code to module
+        self.parts.append(self._get_function(func, indentation_level))
+
+    def _import_module(self, module: str):
+        if module not in (self.target_module, "builtins"):
+            self.imports.add(module)
+
+    def _type_imports(self, type_annotation):
```

Comment:
what's the purpose of this function? it doesn't return anything

Reply:
It's a recursive import tracker for type annotations that add module names to the StubEmitters `self.imports` as we come across them, so they can be added at the top of the type stub file when source is emitted.

I am not a huge fan of having this be such a side-effecty thing, but the code is a lot more managable this way instead of having all different methods be basically "pure" and manage sets of imports they happen to touch upon and return them.

Reply:
oh I see now. If it has side effects, maybe name it something that indicates that (e.g. `_register_import` or similar)